src/Corax/Queries/SortingMatches/SortingMultiMatch.Helpers.cs

```diff
@@ -0,0 +1,111 @@
+
+        return (orderMetadata[2].FieldType, orderMetadata[2].Ascending) switch
+        {
+            (MatchCompareFieldType.Integer, true) => &Fill<TComparer1, TComparer2, EntryComparerByLong>,
+            (MatchCompareFieldType.Integer, false) => &Fill<TComparer1, TComparer2, Descending<EntryComparerByLong>>,
+            (MatchCompareFieldType.Floating, true) => &Fill<TComparer1, TComparer2, EntryComparerByDouble>,
+            (MatchCompareFieldType.Floating, false) => &Fill<TComparer1, TComparer2, Descending<EntryComparerByDouble>>,
+            (MatchCompareFieldType.Sequence, true) => &Fill<TComparer1, TComparer2, EntryComparerByTerm>,
+            (MatchCompareFieldType.Sequence, false) => &Fill<TComparer1, TComparer2, Descending<EntryComparerByTerm>>,
+            (MatchCompareFieldType.Spatial, true) => &Fill<TComparer1, TComparer2, EntryComparerBySpatial>,
+            (MatchCompareFieldType.Spatial, false) => &Fill<TComparer1, TComparer2, Descending<EntryComparerBySpatial>>,
+            (MatchCompareFieldType.Alphanumeric, true) => &Fill<TComparer1, TComparer2, EntryComparerByTermAlphaNumeric>,
+            (MatchCompareFieldType.Alphanumeric, false) => &Fill<TComparer1, TComparer2, Descending<EntryComparerByTermAlphaNumeric>>,
+            _ => &Fill<TComparer1, TComparer2, NullComparer>
+        };
+    }
+    
+    private struct EntryComparerHelper
+    {
+        public static Span<int> NumericSortBatch<TComparer1, TComparer2, TComparer3>(ref SortingMultiMatch<TInner> match, Span<long> batchTermIds, UnmanagedSpan* batchTerms, TComparer1 comparer1, TComparer2 comparer2, TComparer3 comparer3)
+            where TComparer1 : struct, IComparer<UnmanagedSpan>, IComparer<int>, IEntryComparer
+            where TComparer2 : struct, IComparer<UnmanagedSpan>, IComparer<int>, IEntryComparer
+            where TComparer3 : struct, IComparer<UnmanagedSpan>, IComparer<int>, IEntryComparer
+        {
+            var indexes = MemoryMarshal.Cast<long, int>(batchTermIds)[..(batchTermIds.Length)];
+            for (int i = 0; i < batchTermIds.Length; i++)
+            {
+                batchTerms[i] = new UnmanagedSpan(batchTermIds[i]);
+                indexes[i] = i;
+            }
+
+            IndirectSort(ref match, indexes, batchTerms, comparer1, comparer2, comparer3);
+
+            return indexes;
+        }
+
+        public static void IndirectSort<TComparer1, TComparer2, TComparer3>(ref SortingMultiMatch<TInner> match, Span<int> indexes,
+            UnmanagedSpan* batchTerms, TComparer1 comparer1, TComparer2 comparer2, TComparer3 comparer3)
+            where TComparer1 : struct, IComparer<UnmanagedSpan>, IEntryComparer
+            where TComparer2 : struct, IComparer<UnmanagedSpan>, IComparer<int>, IEntryComparer
+            where TComparer3 : struct, IComparer<UnmanagedSpan>, IComparer<int>, IEntryComparer
+        {
+            
+            if (match._orderMetadata[0].Ascending)
```

Comment:
why is this needed?

Reply:
@ayende is it fine then?

Reply:
I'm pretty sure that we can't change the direction of the order, so why not check directly?

Reply:
Well, this is an architectural issue from copying `SortingMatch` since it had to support `SortBatch` and `SortUsingIndex`.

For the first method, the generic type is only for creating the correct `IEntryComparer` to call `SortBatch`, Inside this we create 'new' instance of comparer (e.g. we completely change the comparer (the type) for `CompactKey` in TermComparer). 
This is why we had to wrap it in `IndirectComparer`.

However, in the case of `SortUsingIndex` we use that type immediately and use `Descending<TCmp>`. I thought we will work on `SortUsingIndex` in compound sorting soon so I have not changed the design to avoid additional work that will be removed anyway.